### DISCUSSION TITLE
When I use Env in the class injected by the provider, the build fails

### DISCUSSION THREAD

**Discussion by FrankFSM** [adonisjs/core#1442] [Help] [answered]
```
//provider
public async register () {
    this.container.singleton('Adonis/App/AuthService',()=>{
      return new AuthService()
    });
  }
 
//AuthService
import Env from '@ioc:Adonis/Core/Env'
```
There will be an error like the above
ErrorMessage：IocLookupException: E_IOC_LOOKUP_FAILED: Cannot resolve Adonis/Core/Env namespace from the IoC Container

My solution：
```
//provider
 this.container.singleton('Adonis/App/AuthService',()=>{
      return new AuthService(this.container.use('Adonis/Core/Env'))
    });
//AuthService
import {Env} from '@adonisjs/env/build/src/Env'

  constructor(private env: Env) {
  }
```
I can solve it through the constructor, but is there a better way?

**Comment by thetutlage** [comment, marked as answer]
Well, the entire point of Service providers and Dependency injection is to not have hard dependencies `import statements`. If you don't want to use DI, then why are you even creating a service provider at first place?

**Comment by FrankFSM** [comment]
Thanks. I get it ,I didn't understand the meaning of provider
